I26 vaccination (#29)
Como-DTC-Collaboration/como-models: vignettes/SEIRDV.Rmd

---
title: "The SEIRDV model"
author: "Ioana Bouros"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{The SEIRV model}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(comomodels)
library(ggplot2)
library(glue)
library(dplyr)
```

## Introduction

> This document explains the basics of the SEIR model. The model describes

Reviewer: ```suggestion
This document explains the basics of the SEIRV model. The model describes
```

how populations of susceptible, exposed, infectious, recovered and vaccinated
individuals evolve over time. Waning immunity is also considered in this model,

> so there is a constant supply of susceptibles in the population and make the model

Reviewer: ```suggestion
so there is a constant supply of susceptibles in the population which makes the model
```

more realistic.

## The SEIRDV Model
The SEIRDV model consists of five ODEs describing five populations of susceptible,
recovered, infectious, recovered and vaccinated individuals. Susceptible
individuals ($S$) have not been exposed to the virus. Exposed individuals ($E$) have been
exposed to the virus and have been infected but are not yet infectious to others; they are 
in the incubation period. Infectious individuals ($I$) can spread the virus to 
others.  Recovered individuals ($R$) are no longer
infectious, but they can return to the susceptible population when they lose their

> immunity. Vaccinated individuals ($R$) cannot become infected and are not ill at

Reviewer: ```suggestion
immunity. Vaccinated individuals ($V$) cannot become infected and are not ill at
```

the time, but as it was the case with the recovered population, their immunity

> is temporarily and can become susceptible again with a certain rate. Births and

Reviewer: ```suggestion
is temporary and they can become susceptible again with a certain rate. Births and
```

deaths unrelated to the infection are not considered in this model. The ODEs describing
how the five populations evolve over time are
$$\frac{\text{d}S}{\text{d}t} = -\beta S I - \nu S + \delta_V V + \delta_R R,$$
$$\frac{\text{d}E}{\text{d}t} = \beta S I -\kappa E,$$
$$\frac{\text{d}I}{\text{d}t} = \kappa E - (\gamma + \mu) I,$$
$$\frac{\text{d}R}{\text{d}t} = \gamma I - \delta_R R,$$
$$\frac{\text{d}V}{\text{d}t} = \nu S - \delta_V V$$
where $\beta$, $\kappa$, $\gamma$, $\mu$, $\nu$, $\delta_V$ and $\delta_R$ are
positive parameters. The rate at which susceptible individuals are infected
depends on the fraction of the population that is susceptible, the fraction of
the population that is infectious, and the probability of an infectious individual

> infecting a susceptible individual, $\beta$. This is represented in the

Reviewer: I doubt this is a precise definition for beta, given that in our model the parameter must also account for the average number of contacts, and we happily allow values of beta greater than 1.

Author: I have adapted Solveig's and Hui Jia's vignette to be honest. I agree with your point, so I replaced probability by odds (as in odds ratio), which to me this seems more or less what it is. I have done the same in the SEIRD.Rmd for consistency.

first terms on the right-hand side of $\text{d}S/ \text{d}t$ and $\text{d}E/ \text{d}t$.
Exposed individuals become infectious at rate $\kappa$. $1/ \kappa$ is
the incubation time of the virus, i.e. how many days after exposure the individual
becomes infectious to others. Infectious individuals can either recover or die
because of the disease. Infectious individuals recover in $1/ \gamma$
days, thus individuals recover at rate $\gamma$. Similarly, infectious
individuals die due to the infection after $1/ \mu$ days, making the death rate $\mu$.

> Individuals are vaccinated at rate $\nu$. Among those who are immune (

Reviewer: ```suggestion
Susceptible individuals are vaccinated at rate $\nu$. Among those who are immune (
```

> Individuals are vaccinated at rate $\nu$. Among those who are immune (
> recovered and vaccinated), people lose their immunity against the virus at different

Reviewer: ```suggestion
Individuals are vaccinated at rate $\nu$. Among those who are immune
(recovered and vaccinated), people lose their immunity against the virus at different
```
Apparently, the new lines get turned into spaces

> recovered and vaccinated), people lose their immunity against the virus at different

Reviewer: ```suggestion
recovered and vaccinated), people lose their immunity against the virus (and return to the susceptible compartment) at different
```

rates depending on what type of immunity they acquired: $\delta_V$ for those vaccinated
and $\delta_R$ respectively for those previously infected.

In addition to these four main ODEs, the model also keeps track of the cumulative number 
of cases ($C$) and the cumulative number of disease-related deaths ($D$) over time
$$\frac{\text{d}C}{\text{d}t} = \beta S I,$$
$$\frac{\text{d}D}{\text{d}t} = \mu I.$$

The system is closed by specifying the initial conditions
$$S(0) = S_0,\ E(0) = E_0,\ I(0) = I_0,\ R(0) = R_0,\ V(0) = V_0,\ C(0) = C_0,\ D(0) = D_0,$$

> where $C_0 = E_0 + I_0 + R_0$ to ensure the cases that occur before $t=0$ are

Reviewer: While this sounds sensible, it doesn't seem to match what we are doing in the code, which is always C_0=0

Author: I can just take that piece out to be honest, as we never quite give C0 and D0

We next illustrate how this model works using functionality available in `comomodels`.

## The `SEIRDV` class

We first create an `SEIRDV` object that we will use to run simulations.
```{r}
my_model <- SEIRDV()
```

Next, we set the parameter values for $\beta$, $\kappa$, $\gamma$, $\mu$, $\nu$,
$\delta_V$ and $\delta_R$:
```{r}
params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=0.1,
               delta_R=0.05)
transmission_parameters(my_model) <- params
```

To check that the parameters have been set correctly, we call
```{r}
transmission_parameters(my_model)
```

We now need to set the initial conditions for the model. We only set the initial conditions $S(0)$, $E(0)$, $I(0)$, $R(0)$ and $V(0)$.
```{r}
initial_conditions(my_model) <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
```
The initial conditions must sum to one. If they do not, an error is thrown.

Now we simulate the system from $t=0$ days to $t=50$ days.
```{r}
times <- seq(0, 50, by = 1)
out_df <- run(my_model, times)
```

> The simulation returns two objects: one is a data frame comprising the states over time. Here, there's a peak in the infectious proportion between $t=15$ and $t=20$.

Reviewer: With the current parameter values, it looks like I declines from t=0 to t=10 and is low afterwards.

Author: Ups, I should have actually looked a bit more into what was written there😂

```{r, fig.height = 5, fig.width = 7}
states <- out_df$states

ggplot(states, aes(x = time, y = value)) +
  geom_line(aes(color = compartment)) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population",

>        title = glue("Evolution of different compartments over time"))

Reviewer: It's hard to see E, I, R, and D on this figure because S and V are so much larger. Maybe have two panels, one for each set

number of deaths per day over time.
```{r, fig.height = 5, fig.width = 7}
# function to setup and run model for different beta values
run_seird <- function(delta_V_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=delta_V_val,
                 delta_R=0.05)
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params)
  times <- seq(0, 50, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

# run model across different beta values
delta_V_vals <- c(0.1, 0.05, 0.01, 0.2, 0.5)
for(i in seq_along(delta_V_vals)) {
  delta_V_temp <- delta_V_vals[i]
  temp <- run_seird(delta_V_temp) %>% mutate(delta_V=delta_V_temp)
  if(i == 1)
    result <- temp %>% mutate(delta_V=delta_V_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(delta_V))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  labs(x = "time (days)", y = "Fraction of the population that died per day",

>        title = glue("Number of deaths for differing values of beta")) +

Reviewer: ```suggestion
       title = glue("Number of deaths for differing values of δ_V")) +
```

We now investigate the sensitivity of the number of deaths to the vaccination rate, $\nu$. We run the model for varying values of $\nu$ and plot the
number of deaths per day over time.
```{r, fig.height = 5, fig.width = 7}
# function to setup and run model for different nu values
run_seird <- function(nu_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=nu_val, delta_V=0.1,
                 delta_R=0.05)
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params)
  times <- seq(0, 50, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

# run model across different nu values
nu_vals <- c(0.1, 0.25, 0.5, 0.75, 1)
for(i in seq_along(nu_vals)) {
  nu_temp <- nu_vals[i]
  temp <- run_seird(nu_temp) %>% mutate(nu=nu_temp)
  if(i == 1)
    result <- temp %>% mutate(nu=nu_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(nu))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  labs(x = "time (days)", y = "Fraction of the population that died per day",

>        title = glue("Number of deaths for differing values of beta")) +

Reviewer: ```suggestion
       title = glue("Number of deaths for differing values of ν")) +
```

We next illustrate how this model works using functionality available in `comomodels`.

## The `SEIRDV` class

We first create an `SEIRDV` object that we will use to run simulations.
```{r}
my_model <- SEIRDV()
```

Next, we set the parameter values for $\beta$, $\kappa$, $\gamma$, $\mu$, $\nu$,
$\delta_V$ and $\delta_R$:
```{r}
params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=0.1,
               delta_R=0.05)
transmission_parameters(my_model) <- params
```

To check that the parameters have been set correctly, we call
```{r}
transmission_parameters(my_model)
```

We now need to set the initial conditions for the model. We only set the initial conditions $S(0)$, $E(0)$, $I(0)$, $R(0)$ and $V(0)$.
```{r}
initial_conditions(my_model) <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
```
The initial conditions must sum to one. If they do not, an error is thrown.

Now we simulate the system from $t=0$ days to $t=50$ days.
```{r}
times <- seq(0, 50, by = 1)
out_df <- run(my_model, times)
```

> The simulation returns two objects: one is a data frame comprising the states over time. Here, there's a peak in the infectious proportion close to $t=10$ with the number declining and remaining low afterwards.

Reviewer: Isn't the t=10 peak actually occurring in the R compartment

more realistic.

## The SEIRDV Model
The SEIRDV model consists of five ODEs describing five populations of susceptible,
recovered, infectious, recovered and vaccinated individuals. Susceptible
individuals ($S$) have not been exposed to the virus. Exposed individuals ($E$) have been
exposed to the virus and have been infected but are not yet infectious to others; they are 
in the incubation period. Infectious individuals ($I$) can spread the virus to 
others.  Recovered individuals ($R$) are no longer
infectious, but they can return to the susceptible population when they lose their
immunity. Vaccinated individuals ($V$) cannot become infected and are not ill at
the time, but as it was the case with the recovered population, their immunity
is temporary and they can become susceptible again with a certain rate. 
We can vaccinate not only the susceptibles, but also $E$, $I$ and $R$s. However,
they have already been exposed to the illness so they do not change compartment
assignment. Births and deaths unrelated to the infection are not considered in
this model. The ODEs describing how the five populations evolve over time are

> $$\frac{\text{d}S}{\text{d}t} = -\beta S I - \nu S + \delta_V V + \delta_R R,$$

Reviewer: These equations should now include the intervention

> $$\frac{\text{d}S}{\text{d}t} = -\beta S I - \nu S + \delta_V V + \delta_R R,$$

Reviewer: Not sure if this is something that was considered---is there any point to nu anymore? why not just have the rate of vaccination be one of the intervention parameters directly instead of the 0--1 coverage?
If you keep nu it should be redefined as a maximum possible vaccination rate.

Author: I will do exactly that! Thanks for the suggestion

positive parameters. The rate at which susceptible individuals are infected
depends on the fraction of the population that is susceptible, the fraction of
the population that is infectious, and the odds of an infectious individual
infecting a susceptible individual, $\beta$. This is represented in the
first terms on the right-hand side of $\text{d}S/ \text{d}t$ and $\text{d}E/ \text{d}t$.
Exposed individuals become infectious at rate $\kappa$. $1/ \kappa$ is
the incubation time of the virus, i.e. how many days after exposure the individual
becomes infectious to others. Infectious individuals can either recover or die
because of the disease. Infectious individuals recover in $1/ \gamma$
days, thus individuals recover at rate $\gamma$. Similarly, infectious
individuals die due to the infection after $1/ \mu$ days, making the death rate $\mu$.
Susceptible individuals are vaccinated at rate $\nu$. Among those who are immune
(recovered and vaccinated), people lose their immunity against the virus at different
rates depending on what type of immunity they acquired: $\delta_V$ for those vaccinated
and $\delta_R$ respectively for those previously infected.

> In addition to these four main ODEs, the model also keeps track of the cumulative number 

Reviewer 2: five main ODEs

We next illustrate how this model works using functionality available in `comomodels`.

## The `SEIRDV` class

We first create an `SEIRDV` object that we will use to run simulations.
```{r}
my_model <- SEIRDV()
```

Next, we set the parameter values for $\beta$, $\kappa$, $\gamma$, $\mu$, $\nu$,
$\delta_V$ and $\delta_R$:
```{r}
params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=0.1,
               delta_R=0.05)
transmission_parameters(my_model) <- params
```

To check that the parameters have been set correctly, we call
```{r}
transmission_parameters(my_model)
```

We now need to set the initial conditions for the model. We only set the initial conditions $S(0)$, $E(0)$, $I(0)$, $R(0)$ and $V(0)$.
```{r}
initial_conditions(my_model) <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
```
The initial conditions must sum to one. If they do not, an error is thrown.

Now we simulate the system from $t=0$ days to $t=50$ days.
```{r}
intervention_parameters(my_model) <- list(starts=c(17, 35, 42),
                                          stops=c(25, 39, 49),

>                                           coverages=c(1, 1, 1))

Reviewer: I would add a subsection to the vignette explaining the intervention feature, how it can be used to supply step function vaccination rates, and why these are more realistic than a constant vaccination rate for all time. You could also add a plot of the intervention protocol resulting from these parameter choices.

times <- seq(0, 50, by = 1)
out_df <- run(my_model, times)
```

> The simulation returns two objects: one is a data frame comprising the states over time. Here, the infectious proportion is steadily declining, plummeting to 0 after $t=20$.

Reviewer 2: Is it worth to extend the simulation period longer, e.g. 200? In the first plot for `S` and `V` over time, we can see fluctuations in both states throughout the simulation period, and the R population isn't at steady state yet, thus the system is not stabilized at the end of t=50. 

When increasing the simulation period to 200, the "S-V fluctuations" began "disappear" from around t=100, following a second population boost in both S and R, along with small increase in E and I, which then declined to steady state. It looks interesting - I don't remember if you've already show the phenomenon in the group meeting, just ignore me if you did already :)

ggplot(eird, aes(x = time, y = value)) +
  geom_line(aes(color = compartment)) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population") +
  theme(text = element_text(size = 20))    
```

The simulation also outputs incidence and deaths. These are derived from the state variables. Incidence is just the difference between the cumulative number of cases between two time points,

$$ \text{incidence(t)} = C(t) - C(t - 1),$$

since this shows the number of cases which have arisen between these time  points. Deaths are also given by a difference, but, in this instance, between the cumulative number of deaths at two consecutive time points:

$$ \text{deaths(t)} = D(t) - D(t - 1).$$

```{r fig2, fig.height = 5, fig.width = 7}
changes <- out_df$changes

ggplot(changes, aes(x = time, y = value, fill = compartment)) +
  geom_bar(stat="identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population per day") +
  theme(legend.position = "bottom", legend.title = element_blank(),
        text = element_text(size = 20))

```

> ## Sensitivity analysis

Reviewer 2: It's a quite nice layout of plots to show sensitivity. Would it be useful to write a few lines after each plots as a brief summary for the sensitivity of parameter values and the effects of the values on the epidemic model?

# function to setup and run model for different beta values
run_seirdv <- function(delta_V_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=delta_V_val,
                 delta_R=0.05)
  interv<- list(starts=c(17, 35, 42),
                stops=c(25, 39, 49),
                coverages=c(1, 1, 1))
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params,
                intervention_parameters = interv)
  times <- seq(0, 50, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

# run model across different beta values
delta_V_vals <- c(0.1, 0.05, 0.01, 0.2, 0.5)
for(i in seq_along(delta_V_vals)) {
  delta_V_temp <- delta_V_vals[i]
  temp <- run_seirdv(delta_V_temp) %>% mutate(delta_V=delta_V_temp)
  if(i == 1)
    result <- temp %>% mutate(delta_V=delta_V_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(delta_V))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +

>   labs(x = "time (days)", y = "fraction of the population that died per day") +

Reviewer: ```suggestion
  labs(x = "time (days)", y = "fraction of the population \n that died per day") +
```
and check the other plots as well as some labels are not fitting

  interv<- list(starts=c(17, 35, 42),
                stops=c(25, 39, 49),
                coverages=c(1, 1, 1))
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params,
                intervention_parameters = interv)
  times <- seq(0, 50, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

# run model across different nu values
nu_vals <- c(0.1, 0.25, 0.5, 0.75, 1)
for(i in seq_along(nu_vals)) {
  nu_temp <- nu_vals[i]
  temp <- run_seirdv(nu_temp) %>% mutate(nu=nu_temp)
  if(i == 1)
    result <- temp %>% mutate(nu=nu_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(nu))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population that died per day") +
  theme(legend.position = "bottom", legend.title = element_blank(),
        text=element_text(size=20))

```
We now investigate the sensitivity of the number of deaths to the vaccination rate, $\nu$. We run the model for varying values of $\nu$ and plot the

> number of deaths per day over time.

Reviewer 2: This section investigates different `start_val` rather than `nu`. Same for the commented lines below.

  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population") +
  theme(text = element_text(size = 20))    

eird <- subset(states, !states$compartment %in% c("S", "V"))
ggplot(eird, aes(x = time, y = value)) +
  geom_line(aes(color = compartment)) +
  scale_color_brewer(palette = "Dark2") +
  labs(x = "time (days)", y = "fraction of the population") +
  theme(text = element_text(size = 20))    
```

The simulation also outputs incidence and deaths. These are derived from the state variables. Incidence is just the difference between the cumulative number of cases between two time points,

$$ \text{incidence(t)} = C(t) - C(t - 1),$$

since this shows the number of cases which have arisen between these time  points. Deaths are also given by a difference, but, in this instance, between the cumulative number of deaths at two consecutive time points:

$$ \text{deaths(t)} = D(t) - D(t - 1).$$

```{r fig2, fig.height = 5, fig.width = 7}
changes <- out_df$changes

ggplot(changes, aes(x = time, y = value, fill = compartment)) +
  geom_bar(stat="identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +

>   labs(x = "time (days)", y = "fraction of the population per day") +

Reviewer: ```suggestion
  labs(x = "time (days)", y = "fraction of the population \n per day") +
```

  theme(legend.position = "bottom", legend.title = element_blank(),
        text = element_text(size = 20))

```

## Sensitivity analysis
In this section, we investigate the sensitivity of the number of deaths to the loss-of-immunity rate
in vaccinated individuals, $\delta_V$. We run the model for varying values of $\delta_V$ and plot the
number of deaths per day over time.
```{r, fig.height = 5, fig.width = 7}

> # function to setup and run model for different beta values

Reviewer: ```suggestion
# function to setup and run model for different delta_V values
```

run_seirdv <- function(delta_V_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=delta_V_val,
                 delta_R=0.05)
  interv<- list(starts=c(17, 35, 42),
                stops=c(25, 39, 49),
                coverages=c(1, 1, 1))
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params,
                intervention_parameters = interv)
  times <- seq(0, 200, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

> # run model across different beta values

Reviewer: ```suggestion
# run model across different delta_V values
```

delta_V_vals <- c(0.1, 0.05, 0.01, 0.2, 0.5)
for(i in seq_along(delta_V_vals)) {
  delta_V_temp <- delta_V_vals[i]
  temp <- run_seirdv(delta_V_temp) %>% mutate(delta_V=delta_V_temp)
  if(i == 1)
    result <- temp %>% mutate(delta_V=delta_V_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(delta_V))) +

>   geom_bar(stat = "identity", position = position_dodge()) +

Reviewer: ```suggestion
  ggplot(aes(x=time, y=value, color = as.factor(delta_V))) +
  geom_line(stat = "identity", position = position_dodge()) +
```
What do you think of using line plots for these
I know we've used bars previously, but for this figure and the two below, I would say that the results are way clearer with lines.

number of deaths per day over time.
```{r, fig.height = 5, fig.width = 7}
# function to setup and run model for different nu values
run_seirdv <- function(nu_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=nu_val, delta_V=0.1,
                 delta_R=0.05)
  interv<- list(starts=c(17, 35, 42),
                stops=c(25, 39, 49),
                coverages=c(1, 1, 1))
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params,
                intervention_parameters = interv)
  times <- seq(0, 200, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

# run model across different nu values
nu_vals <- c(0.1, 0.25, 0.5, 0.75, 1)
for(i in seq_along(nu_vals)) {
  nu_temp <- nu_vals[i]
  temp <- run_seirdv(nu_temp) %>% mutate(nu=nu_temp)
  if(i == 1)
    result <- temp %>% mutate(nu=nu_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(nu))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +

>   labs(x = "time (days)", y = "fraction of the population that died per day") +

Reviewer: ```suggestion
  labs(x = "time (days)", y = "fraction of the population \n that died per day") +
```

  theme(legend.position = "bottom", legend.title = element_blank(),
        text=element_text(size=20))

```
As we can see from the plot above, larger rates of maximum vaccination $\nu$ make the second wave of the number of deaths in the epidemic occur much further to the first peak. For larger $\nu$, we see no deaths for larger stretches of time, even though the second peak is larger.

> We now investigate the sensitivity of the start time of intervention values, `start_val`. We run the model for varying values of `start_val` and plot the

Reviewer: Mention that for this investigation, we are switching to a new intervention that starts at start_val and continues to 100, rather than the one you've been using previously

number of deaths per day over time.
```{r, fig.height = 5, fig.width = 7}
# function to setup and run model for different start intervention values
run_seirdv <- function(start_val) {
  inits <- list(S0=0.99, E0=0, I0=0.01, R0=0, V0=0)
  params <- list(beta=1, kappa=0.9, gamma=0.5, mu=0.1, nu=0.4, delta_V=0.1,
                 delta_R=0.05)
  interv<- list(starts=start_val,
                stops=100,
                coverages=1)
  model <- SEIRDV(initial_conditions = inits,
                transmission_parameters=params,
                intervention_parameters = interv)
  times <- seq(0, 200, by = 1)
  out_df <- run(model, times)
  out_df$changes
}

> # run model across different nu values

Reviewer: ```suggestion
# run model across different start_val values
```

start_val <- c(0, 5, 10, 15, 20, 30, 45, 55)
for(i in seq_along(start_val)) {
  start_temp <- start_val[i]
  temp <- run_seirdv(start_temp) %>% mutate(starts=start_temp)
  if(i == 1)
    result <- temp %>% mutate(starts=start_temp)
  else
    result <- result %>% bind_rows(temp)
}

# plot results
result %>% 
  filter(compartment=="Deaths") %>% 
  ggplot(aes(x=time, y=value, fill = as.factor(starts))) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_color_brewer(palette = "Dark2") +

>   labs(x = "time (days)", y = "fraction of the population that died per day") +

Reviewer: ```suggestion
  labs(x = "time (days)", y = "fraction of the population \n that died per day") +
```

  theme(legend.position = "bottom", legend.title = element_blank(),
        text=element_text(size=20))

```

> As we can see from the plot above, larger start time of intervention values, `start_val` produces deaths for much longer and the peak of their number is higher than in thr case of smaller `start_val`.

Reviewer: ```suggestion
As we can see from the plot above, larger start time of intervention values, `start_val` produces deaths for much longer and the peak of their number is higher than in the case of smaller `start_val`.
```

more realistic.

## The SEIRDV Model
The SEIRDV model consists of five ODEs describing five populations of susceptible,
recovered, infectious, recovered and vaccinated individuals. Susceptible
individuals ($S$) have not been exposed to the virus. Exposed individuals ($E$) have been
exposed to the virus and have been infected but are not yet infectious to others; they are 
in the incubation period. Infectious individuals ($I$) can spread the virus to 
others.  Recovered individuals ($R$) are no longer
infectious, but they can return to the susceptible population when they lose their
immunity. Vaccinated individuals ($V$) cannot become infected and are not ill at
the time, but as it was the case with the recovered population, their immunity
is temporary and they can become susceptible again with a certain rate. 
We can vaccinate not only the susceptibles, but also $E$, $I$ and $R$s. However,
they have already been exposed to the illness so they do not change compartment
assignment. Births and deaths unrelated to the infection are not considered in
this model. The ODEs describing how the five populations evolve over time are

> $$\frac{\text{d}S}{\text{d}t} = -\beta S I - \nu Inter(t) S + \delta_V V + \delta_R R,$$

Reviewer: ```suggestion
$$\frac{\text{d}S}{\text{d}t} = -\beta S I - \nu \text{Inter}(t) S + \delta_V V + \delta_R R,$$
```
and twice below.
